src/main/java/com/metamx/common/FileUtils.java

```diff
@@ -0,0 +1,302 @@
+  // Child size is eagerly calculated and any modifications to the file after the child is added are not accounted for.
+  // As such, this result and all child results should be considered immutable and should be built from the leafs up.
+  public static class FileCopyResult
+  {
+    private static final int DEFAULT_BUFFER_SIZE = 1 << 10;
+    private final Collection<FileCopyResult> children = Lists.newArrayList();
+    private final Path path;
+    private long size = 0l;
+    private final MessageDigest digest;
+    private final byte[] digestBytes;
+
+    public Collection<FileCopyResult> getChildren()
+    {
+      return ImmutableList.copyOf(children);
+    }
+
+    public MessageDigest getDigest()
+    {
+      return digest;
+    }
+
+    public Path getPath()
+    {
+      return path;
+    }
+
+    // Consumes otherDigest.digest()
+    public Boolean digestEquals(MessageDigest otherDigest)
+    {
+      if (this.digest == null || this.digestBytes == null) {
+        return false;
+      }
+      if (!otherDigest.getAlgorithm().equals(digest.getAlgorithm())) {
+        throw new IAE(
+            "Digest algorithm did not match. Expected [%s] got [%s]",
+            digest.getAlgorithm(),
+            otherDigest.getAlgorithm()
+        );
+      }
+      return Arrays.equals(digestBytes, otherDigest.digest());
+    }
+
+    // Only works for immutable children contents
+    public long size()
+    {
+      return size;
+    }
+
+    // This consumes digest.digest()
+    public FileCopyResult(Path path, MessageDigest digest, FileCopyResult... children)
+    {
+      this.path = path;
+      final File file = path.toFile();
+      if (file.exists() && !file.isDirectory()) {
+        size += file.length();
+      }
+      if (children != null) {
+        this.addSizedChildren(Arrays.asList(children));
+      }
+      this.digest = digest;
+      this.digestBytes = digest.digest();
+    }
+
+    public FileCopyResult(Path path, Collection<FileCopyResult> children)
+    {
+      this.path = path;
+      this.addChildren(children);
+      this.digest = null;
+      this.digestBytes = null;
+    }
+
+    private void addSizedChildren(Collection<FileCopyResult> children)
+    {
+      if (children == null || children.isEmpty()) {
+        return;
+      }
+      long size = 0l;
+      for (FileCopyResult child : children) {
+        size += child.size();
+      }
+      this.children.addAll(children);
+      this.size += size;
+    }
+
+    public void addChildren(Collection<FileCopyResult> children)
+    {
+      this.addSizedChildren(children);
+    }
+
+    public void addChild(FileCopyResult child)
+    {
+      this.addChildren(ImmutableList.of(child));
+    }
+
+    public static FileCopyResult buildResult(Path path, FileCopyResult... children) throws IOException
+    {
+      return buildResult(path, "MD5", children);
+    }
+
+    public static FileCopyResult buildResult(
+        Path path,
+        String digestAlgorithm,
```

Comment:
taking MessageDigest objects instead of just String might make the API more explicit and less confusing as to what algorithms are supported. It would also remove all the unnecessary checks around getting an instance of the digest?